### PR TITLE
add validations

### PR DESCRIPTION
**Pull request by Garvshah01**

### CODE REVIEW

File: app/models/product.rb [hunk 1]
```diff
@@ -1,12 +1,34 @@
+class UrlValidator < ActiveModel::EachValidator
```
Comment (shivamvinsol) [inline app/models/product.rb]: create a new folder `validators` and move it there.

File: app/models/product.rb [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+  def validate_each(record, attr, value)
+    unless value =~ %r{\.(gif|jpg|png)\z}i
```
Comment (shivamvinsol) [inline app/models/product.rb]: move it to a constant.

File: app/models/product.rb [hunk 3]
```diff
@@ -1,12 +1,34 @@
...
+      record.errors.add attr, 'must be a URL for GIF, JPG or PNG images'
+    end
+  end
+end
+
+class PriceValidator < ActiveModel::Validator
```
Comment (shivamvinsol) [inline app/models/product.rb]: move to validators

File: app/models/product.rb [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
+  def validate(record)
+    if record[:price] < record[:discount_price]
+      record.errors.add :price, "should be greater than #{record[:discount_price]}"
+    end
+  end
+end
+
 class Product < ApplicationRecord
 
+  validates :permalink, uniqueness:true
```
Comment (shivamvinsol) [inline app/models/product.rb]: spacing

Comment (shivamvinsol) [reply app/models/product.rb]: allow_nil ?

File: app/models/product.rb [hunk 5]
```diff
@@ -1,12 +1,34 @@
...
+  validates :permalink, format: {
+    with: %r{\A[A-Za-z0-9-]+\Z},
```
Comment (shivamvinsol) [inline app/models/product.rb]: - regex should be constant
- do we need both A-Z and a-z? 
- allow_nil?

File: app/models/product.rb [hunk 6]
```diff
@@ -1,12 +1,34 @@
...
+
+class PriceValidator < ActiveModel::Validator
+  def validate(record)
+    if record[:price] < record[:discount_price]
+      record.errors.add :price, "should be greater than #{record[:discount_price]}"
+    end
+  end
+end
+
 class Product < ApplicationRecord
```
Comment (shivamvinsol) [inline app/models/product.rb:1]: fix spacing issues throughout the code

File: app/models/product.rb [hunk 7]
```diff
@@ -1,12 +1,34 @@
...
 
+  validates :permalink, uniqueness:true
+  validates :permalink, format: {
+    with: %r{\A[A-Za-z0-9-]+\Z},
+    message: 'should not contains any special characters'
+  }
+  validates :words_in_permalink, length: {minimum: 3}
+  validates :words_in_description , length: {in: 5..10}
+  validates :price, comparison: { greater_than: :discount_price }
+  validates_with PriceValidator
```
Comment (shivamvinsol) [inline app/models/product.rb]: move this in end

File: app/models/product.rb [hunk 8]
```diff
@@ -19,5 +41,25 @@ def ensure_not_referenced_by_any_line_item
...
+
+
+
+
+
+
+
+# validates :permalink, format: {
+#   with: %r{\A([A-Za-z0-9]+\-){2,}([A-Za-z0-9]+)\Z}
+# }
```
Comment (shivamvinsol) [inline app/models/product.rb]: remove commented code.

File: app/models/product.rb [hunk 9]
```diff
@@ -19,5 +41,25 @@ def ensure_not_referenced_by_any_line_item
...
+
+
+
+
+
+
+
+
+
+
```
Comment (shivamvinsol) [inline app/models/product.rb]: remove extra line breaks

File: app/models/user.rb [hunk 10]
```diff
@@ -3,6 +3,11 @@ class User < ApplicationRecord
   after_destroy :ensure_an_admin_remains
   validates :name, presence: true, uniqueness: true
   has_secure_password
+  validates :email, uniqueness: true
+  validates :email, format: {
+    with: %r{\A[a-zA-Z0-9+_.-]+@[a-zA-Z0-9.-]+\Z}
```
Comment (shivamvinsol) [inline app/models/user.rb]: regex should be constant.
look for correct email regex from google

Comment (shivamvinsol) [reply app/models/user.rb]: allow_nil ?

File: app/models/product.rb [hunk 11]
```diff
@@ -1,12 +1,34 @@
...
 class Product < ApplicationRecord
 
+  validates :permalink, uniqueness:true
+  validates :permalink, format: {
+    with: %r{\A[A-Za-z0-9-]+\Z},
+    message: 'should not contains any special characters'
+  }
+  validates :words_in_permalink, length: {minimum: 3}
+  validates :words_in_description , length: {in: 5..10}
+  validates :price, comparison: { greater_than: :discount_price }
```
Comment (shivamvinsol) [inline app/models/product.rb]: not equal to?

File: app/models/product.rb [hunk 12]
```diff
@@ -1,12 +1,34 @@
...
+
 class Product < ApplicationRecord
 
+  validates :permalink, uniqueness:true
+  validates :permalink, format: {
+    with: %r{\A[A-Za-z0-9-]+\Z},
+    message: 'should not contains any special characters'
+  }
+  validates :words_in_permalink, length: {minimum: 3}
+  validates :words_in_description , length: {in: 5..10}
```
Comment (shivamvinsol) [inline app/models/product.rb]: we should run validations on model attributes. Lets add custom methods for validating these.

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: app/models/product.rb (hunk 11 above)
Comment (Garvshah01) [reply app/models/product.rb]: Sir price cannot be smaller or equal to discount_price

File: app/models/Validators/validators.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+class UrlValidator < ActiveModel::EachValidator
+  URL_REGEXP = %r{\.(gif|jpg|png)\z}i.freeze
+  def validate_each(record, attr, value)
+    unless value =~ REGEXP
```
Comment (shivamvinsol) [inline app/models/Validators/validators.rb]: how is this working?
where is REGEXP defined?

File: app/models/Validators/validators.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+class UrlValidator < ActiveModel::EachValidator
+  URL_REGEXP = %r{\.(gif|jpg|png)\z}i.freeze
```
Comment (shivamvinsol) [inline app/models/Validators/validators.rb]: add a new line before and after constant.
rename to FILE_TYPE_REGEX

File: app/models/Validators/validators.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  def validate_each(record, attr, value)
+    unless value =~ REGEXP
+      record.errors.add attr, 'must be a URL for GIF, JPG or PNG images'
+    end
+  end
+end
+
+class PriceValidator < ActiveModel::Validator
```
Comment (shivamvinsol) [inline app/models/Validators/validators.rb]: create a separate file for this.

File: app/models/product.rb [hunk 16]
```diff
@@ -1,12 +1,22 @@
+load 'Validators/validators.rb'
```
Comment (shivamvinsol) [inline app/models/product.rb]: why is this needed?

File: app/models/product.rb [hunk 17]
```diff
@@ -1,12 +1,22 @@
...
+
 class Product < ApplicationRecord
 
+  PERMALINK_REGEXP =  %r{\A[a-z0-9-]+\Z}.freeze
```
Comment (shivamvinsol) [inline app/models/product.rb]: capital alphabets are excluded.
can we add some option in regexp?

File: app/models/product.rb [hunk 18]
```diff
@@ -1,12 +1,22 @@
...
-    with: %r{\.(gif|jpg|png)\z}i,
-    message: 'must be a URL for GIF, JPG or PNG images'
+  validates :image_url, allow_blank: true, url: true
+  validates :permalink, format: {
+    with: PERMALINK_REGEXP,
+    message: 'should not contains any special characters'
   }
+  validates :permalink, uniqueness: true, allow_nil: true
+  validate :validate_words_in_permalink
+  validate :validate_words_in_description
```
Comment (shivamvinsol) [inline app/models/product.rb:20]: add custom validations in end

File: app/models/product.rb [hunk 19]
```diff
@@ -1,12 +1,22 @@
...
+  validates :price, numericality: {greater_than_or_equal_to: 0.01}, if: :price
+  validates :price, comparison: { greater_than: :discount_price }
```
Comment (shivamvinsol) [inline app/models/product.rb]: allow_nil??

File: app/models/product.rb [hunk 20]
```diff
@@ -20,4 +30,19 @@ def ensure_not_referenced_by_any_line_item
...
+  def validate_words_in_permalink
+    words = permalink.split('-')
+    if words.size < 3
+      errors.add :permalink, "should not have words less than 3"
+    end
+  end
+
+  def validate_words_in_description
+    words = description.split(' ')
+    if words.size() < 5
```
Comment (shivamvinsol) [inline app/models/product.rb]: do we need ()?

File: app/models/user.rb [hunk 21]
```diff
@@ -1,8 +1,13 @@
 class User < ApplicationRecord
 
+  EMAIL_REGEXP = /^[\w\.-]+@([\w-]+\.)+[\w-]{2,4}$/.freeze
   after_destroy :ensure_an_admin_remains
   validates :name, presence: true, uniqueness: true
   has_secure_password
+  validates :email, uniqueness: true,allow_nil: true
+  validates :email, format: {
```
Comment (shivamvinsol) [inline app/models/user.rb:7]: allow_nil

File: app/models/user.rb [hunk 22]
```diff
@@ -1,8 +1,13 @@
 class User < ApplicationRecord
 
+  EMAIL_REGEXP = /^[\w\.-]+@([\w-]+\.)+[\w-]{2,4}$/.freeze
```
Comment (shivamvinsol) [inline app/models/user.rb]: add a new line

Comment (shivamvinsol) [reply app/models/user.rb]: this is not a valid email regex. check on google

File: app/models/user.rb [hunk 23]
```diff
@@ -1,8 +1,13 @@
...
   after_destroy :ensure_an_admin_remains
   validates :name, presence: true, uniqueness: true
   has_secure_password
+  validates :email, uniqueness: true,allow_nil: true
```
Comment (shivamvinsol) [inline app/models/user.rb]: again spacing issue!!!

File: app/models/product.rb [hunk 24]
```diff
@@ -20,4 +30,19 @@ def ensure_not_referenced_by_any_line_item
     end
   end
 
+  def validate_words_in_permalink
+    words = permalink.split('-')
+    if words.size < 3
```
Comment (shivamvinsol) [inline app/models/product.rb]: MINIMUM_PERMALINK_LENGTH

File: app/models/product.rb (hunk 20 above)
Comment (shivamvinsol) [inline app/models/product.rb]: move 5 to constant

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 21 above)
Comment (Garvshah01) [reply app/models/user.rb:7]: email can be nil

File: app/models/product.rb [hunk 25]
```diff
@@ -1,17 +1,26 @@
...
 
+  PERMALINK_REGEXP =  %r{\A[a-z0-9-]+\Z}.freeze
+
   validates :title, :description, :image_url, presence: true
-  validates :price, numericality: {greater_than_or_equal_to: 0.01}
   validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {
-    with: %r{\.(gif|jpg|png)\z}i,
-    message: 'must be a URL for GIF, JPG or PNG images'
+  validates :image_url, allow_blank: true, url: true
```
Comment (shivamvinsol) [inline app/models/product.rb]: allow_blank should be last option

File: app/models/product.rb [hunk 26]
```diff
@@ -20,4 +29,24 @@ def ensure_not_referenced_by_any_line_item
     end
   end
 
+  MINIMUM_PERMALINK_LENGTH = 3
```
Comment (shivamvinsol) [inline app/models/product.rb]: constants should be a top of class

File: app/models/product.rb [hunk 27]
```diff
@@ -20,4 +29,24 @@ def ensure_not_referenced_by_any_line_item
...
+
+  def validate_words_in_permalink
+    words = permalink.split '-'
+    if words.size < MINIMUM_PERMALINK_LENGTH
+      errors.add :permalink, "should not have words less than 3"
+    end
+  end
+
+  MINIMUM_DESCRIPTION_LENGTH = 5
+  MAXIMUM_DECRIPTION_LENGTH = 10
```
Comment (shivamvinsol) [inline app/models/product.rb]: same ^

File: app/validators/price_validator.rb [hunk 28]
```diff
@@ -0,0 +1,7 @@
+class PriceValidator < ActiveModel::Validator
+  def validate(record)
+    if  record[:price] < record[:discount_price]
```
Comment (shivamvinsol) [inline app/validators/price_validator.rb]: spacing issue

File: app/validators/url_validator.rb [hunk 29]
```diff
@@ -0,0 +1,10 @@
+class UrlValidator < ActiveModel::EachValidator
+
+  URL_REGEXP = %r{\.(gif|jpg|png)\z}i.freeze
+
+  def validate_each(record, attr, value)
+    unless value =~ URL_REGEXP
```
Comment (shivamvinsol) [inline app/validators/url_validator.rb]: use .match?

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: app/validators/url_validator.rb (hunk 29 above)
Comment (Garvshah01) [reply app/validators/url_validator.rb]: done

File: app/validators/price_validator.rb (hunk 28 above)
Comment (Garvshah01) [reply app/validators/price_validator.rb]: done

File: app/models/product.rb (hunk 27 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 26 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 25 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 20 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 24 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/user.rb (hunk 23 above)
Comment (Garvshah01) [reply app/models/user.rb]: done

File: app/models/user.rb (hunk 22 above)
Comment (Garvshah01) [reply app/models/user.rb]: done

File: app/models/user.rb (hunk 21 above)
Comment (Garvshah01) [reply app/models/user.rb:7]: done

File: app/models/product.rb (hunk 17 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 20 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 19 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 18 above)
Comment (Garvshah01) [reply app/models/product.rb:20]: done

File: app/models/product.rb (hunk 16 above)
Comment (Garvshah01) [reply app/models/product.rb]: not needed

File: app/models/Validators/validators.rb (hunk 15 above)
Comment (Garvshah01) [reply app/models/Validators/validators.rb]: done

File: app/models/Validators/validators.rb (hunk 13 above)
Comment (Garvshah01) [reply app/models/Validators/validators.rb]: done

File: app/models/product.rb (hunk 12 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/user.rb (hunk 10 above)
Comment (Garvshah01) [reply app/models/user.rb]: done

File: app/models/product.rb (hunk 1 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 2 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 3 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 4 above)
Comment (Garvshah01) [reply app/models/product.rb]: yes required

File: app/models/product.rb (hunk 5 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 6 above)
Comment (Garvshah01) [reply app/models/product.rb:1]: done

File: app/models/product.rb (hunk 7 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 8 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 9 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb [hunk 30]
```diff
@@ -20,4 +32,19 @@ def ensure_not_referenced_by_any_line_item
...
 
+  def validate_words_in_permalink
+    words = permalink.split '-'
+    if words.size < MINIMUM_PERMALINK_LENGTH
+      errors.add :permalink, "should not have words less than 3"
+    end
+  end
+
+  def validate_words_in_description
+    words = description.split ' '
```
Comment (shivamvinsol) [inline app/models/product.rb]: words_count = description.split(' ').size

File: app/models/product.rb [hunk 31]
```diff
@@ -20,4 +32,19 @@ def ensure_not_referenced_by_any_line_item
     end
   end
 
+  def validate_words_in_permalink
+    words = permalink.split '-'
```
Comment (shivamvinsol) [inline app/models/product.rb]: words_count = description.split(' ').size

File: app/models/product.rb [hunk 32]
```diff
@@ -1,17 +1,29 @@
...
+  validates :price, comparison: { greater_than: :discount_price },  allow_nil: true
+  validates_with PriceValidator, if: :price
+  validate :validate_words_in_permalink
+  validate :validate_words_in_description
 
   has_many :line_items
   has_many :orders, through: :line_items
   before_destroy :ensure_not_referenced_by_any_line_item
 
+
```
Comment (shivamvinsol) [inline app/models/product.rb]: remove

File: app/models/product.rb [hunk 33]
```diff
@@ -1,17 +1,29 @@
 class Product < ApplicationRecord
 
+  PERMALINK_REGEXP =  %r{\A[a-z0-9-]+\Z}i.freeze
+  MINIMUM_PERMALINK_LENGTH = 3
+  MINIMUM_DESCRIPTION_LENGTH = 5
+  MAXIMUM_DECRIPTION_LENGTH = 10
```
Comment (shivamvinsol) [inline app/models/product.rb:6]: freeze these as well

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: app/models/product.rb (hunk 30 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 31 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 32 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 33 above)
Comment (Garvshah01) [reply app/models/product.rb:6]: done

**Review by shivamvinsol — APPROVED** [review]
